### PR TITLE
Add submission/code authority judgement

### PR DESCRIPTION
**Pull request by henryyang42**
### Add submission/code authority judgement

implement the rules stated in [hackpad](https://nthuoj.hackpad.com/Zjv6kKXIj9o#:h=Authorities) using custom template filters

### CODE REVIEW

File: status/templatetags/status_filters.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+    # basic requirement: submission must be shown
+    if show_submission(submission, user):
+        # admin can see everyone's detail
+        if user.user_level == user.ADMIN:
+            return True
+        # no one can see admin's detail
+        if submission.user.user_level == user.ADMIN:
+            return False
+
+        contests = Contest.objects.filter(
```
Comment (henryyang42) [inline status/templatetags/status_filters.py]: Is that filter correct for contest mode?

Comment (nickelcheng) [inline status/templatetags/status_filters.py]: I think it's correct.

File: status/templatetags/status_filters.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+
+    # user's own submission must be seen
+    if submission.user == user:
+        return True
+
+    # problem owner can see all submission of his problem
+    if submission.problem.owner_id == user.username:
+        return True
+
+    # contest owner & coowner's submission can't be seen until the end
```
Comment (nickelcheng) [inline status/templatetags/status_filters.py]: Owner / coowners' submission can be seen if submit time is earlier than contest creation time.

File: status/templatetags/status_filters.py [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+            return True
+        else:
+            # not a owner/coowner
+            # to see submission, submission.user must not be owners
+            return submission.user not in owners
+
+    valid = True
+    # an invisible problem's submission can't be seen
+    if not submission.problem.visible:
+        valid = False
```
Comment (nickelcheng) [inline status/templatetags/status_filters.py]: Why don't you return false directly?

File: status/templatetags/status_filters.py [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+    # problem owner's submission can't be seen
+    if submission.user.username == submission.problem.owner_id:
+        valid = False
```
Comment (nickelcheng) [inline status/templatetags/status_filters.py]: Ditto.

File: status/templatetags/status_filters.py [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+            end_time__gt=datetime.now())
+        # during the contest, only owner/coowner with user level sub-judge/judge 
+        # can view the detail
+        if contests:
+            contests = contests.filter(problem=submission.problem)
+            owners = []
+            for contest in contests:
+                owners.append(contest.owner)
+                owners.extend(contest.coowner.all())
+            if user.has_subjudge_auth() and user in owners:
```
Comment (nickelcheng) [inline status/templatetags/status_filters.py]: An owner or coowner must be sub-judge or judge or admin, so you don't have to check it.

File: status/templates/status/status.html [hunk 6]
```diff
@@ -23,6 +23,7 @@
 -->
 {% extends "index/base.html" %}
 {% load static %}
+{% load status_filters %}
 
 {% block body_block %}
 {% regroup submission_details by sid as submissions %}
```
Comment (nickelcheng) [inline status/templates/status/status.html]: Only All accepted and Not accepted submissions have submission detail.

**Comment by nickelcheng** [comment]
This PR is ok for me. @CHIAHSINHOU please test this.

**Comment by CHIAHSINHOU** [comment]
It's ok.

**Comment by bruce3557** [comment]
Logic is clear, +1 and going to merge

**Pull request merged**